nifi-nar-bundles/nifi-box-bundle/nifi-box-processors/src/main/java/org/apache/nifi/processors/box/PutBoxFile.java

```diff
@@ -0,0 +1,366 @@
+    }
+
+    private BoxFolder.Info getOrCreateParentSubfolder(String folderName, String parentFolderId, boolean createFolder)  {
+        final int indexOfPathSeparator = folderName.indexOf("/");
+
+        if (isMultiLevelFolder(indexOfPathSeparator, folderName)) {
+            final String mainFolderName = folderName.substring(0, indexOfPathSeparator);
+            final String subFolders = folderName.substring(indexOfPathSeparator + 1);
+            final BoxFolder.Info mainFolder = getOrCreateFolder(mainFolderName, parentFolderId, createFolder);
+            return getOrCreateParentSubfolder(subFolders, mainFolder.getID(), createFolder);
+        } else {
+            return getOrCreateFolder(folderName, parentFolderId, createFolder);
+        }
+    }
+
+    private BoxFolder.Info getOrCreateFolder(String folderName, String parentFolderId, boolean createFolder) {
+        final Optional<BoxFolder.Info> existingFolder = checkFolderExistence(folderName, parentFolderId);
+
+        if (existingFolder.isPresent()) {
+            return existingFolder.get();
+        }
+
+        if (createFolder) {
+            getLogger().debug("Create folder " + folderName + " parent id: " + parentFolderId);
+
+            final BoxFolder parentFolder = getFolder(parentFolderId);
+            return parentFolder.createFolder(folderName);
+        } else {
+            throw new ProcessException(format("The specified subfolder '%s' does not exist and '%s' is false.", folderName, CREATE_SUBFOLDER.getDisplayName()));
+        }
+    }
+
+    private Optional<BoxFolder.Info> checkFolderExistence(final String folderName, final String parentFolderId) {
+        final BoxFolder parentFolder = getFolder(parentFolderId);
+        return StreamSupport.stream(parentFolder.getChildren("name").spliterator(), false)
+                .filter(BoxFolder.Info.class::isInstance)
+                .map(BoxFolder.Info.class::cast)
+                .filter(info -> info.getName().equals(folderName))
+                .findAny();
+    }
+
+    private Optional<BoxFile.Info> getFileByName(final String filename, final BoxFolder parentFolder) {
+        return StreamSupport.stream(parentFolder.getChildren("name").spliterator(), false)
+                .filter(BoxFile.Info.class::isInstance)
+                .map(BoxFile.Info.class::cast)
+                .filter(info -> info.getName().equals(filename))
+                .findAny();
+    }
+
+    private boolean isMultiLevelFolder(int indexOfPathSeparator, String folderName) {
+        return indexOfPathSeparator > 0 && indexOfPathSeparator < folderName.length() - 1;
+    }
+
+    private void handleUploadError(final String conflictResolution, final String filename, BoxFolder folder, final BoxAPIException e) {
+        if (e.getResponseCode() == CONFLICT_RESPONSE_CODE) {
+            handleConflict(conflictResolution, filename, folder, e);
+        } else {
+            throw new ProcessException(e);
```

Comment:
Recommend including a specific message along with the wrapped exception.
```suggestion
            throw new ProcessException(String.format("File [%s] upload failed", filename), e);
```

Reply:
I would not wrap the exception here at all, just throw the original one. This way it would land in the right `catch` block of the outer `try-catch` where `BoxAPIResponseException`-s are handled.
It may be weird to rethrow an exception that was passed in as a parameter though. For this reason, it may be better to move this `if-else` back to the inner catch block instead of the separate method.
```
            } catch (BoxAPIResponseException e) {
                if (e.getResponseCode() == CONFLICT_RESPONSE_CODE) {
                    handleConflict(conflictResolution, filename, parentFolder, e);
                } else {
                    throw e;
                }
            }

            session.transfer(flowFile, REL_SUCCESS);
        } catch (BoxAPIResponseException e) {
            getLogger().error("Exception occurred while uploading file '{}' to Box folder '{}'", filename, fullPath, e);
            handleExpectedError(session, flowFile, e);
        }
```
I think this way it is more straightforward which `BoxAPIResponseException` is handled specially and which go to the common handler block.